hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestStructuredStreaming.scala

```diff
@@ -50,6 +51,10 @@ class TestStructuredStreaming extends HoodieClientTestBase {
     DataSourceWriteOptions.RECORDKEY_FIELD.key -> "_row_key",
     DataSourceWriteOptions.PARTITIONPATH_FIELD.key -> "partition",
     DataSourceWriteOptions.PRECOMBINE_FIELD.key -> "timestamp",
+    DataSourceWriteOptions.KEYGENERATOR_CLASS_NAME.key -> classOf[ComplexKeyGenerator].getName,
+    DataSourceWriteOptions.OPERATION.key -> DataSourceWriteOptions.BULK_INSERT_OPERATION_OPT_VAL,
+    // HoodieWriteConfig.BULK_INSERT_WRITE_STREAM_ENABLE.key -> "true",
```

Comment:
does tests need more fixes?

Reply:
this should not be part of the PR. i'll remove it
For the unit test itself, I tried but it's hard. In an attempt to simulate the scenario, the whole test gets killed. However, I have manually verified the fix.

Reply:
I inject a configurable error throw in the own project to verify the rare failover case. Not sure it is suitable way here, as the error injection makes code less readable and ugly.